File: M3Tools/Controls/Products/ProductsComboBox.vb
Imports System.ComponentModel

Public Class ProductsComboBox
	Public Event LoadBegin()
	Public Event LoadEnd()

	Public Property SelectedItem As Object
		Get
			Return cbx_Items.SelectedItem
		End Get
		Set(value As Object)
			If value Is Nothing Then
				Return
			End If

			cbx_Items.SelectedItem = value
		End Set
	End Property

	Public Property SelectedIndex As Integer
		Get
			Return cbx_Items.SelectedIndex
		End Get
		Set(value As Integer)
			If value < 0 Then
				Return
			End If

			cbx_Items.SelectedIndex = value
		End Set
	End Property

	Public Property SelectedValue As Object
		Get
			Return cbx_Items.SelectedValue
		End Get
		Set(value As Object)
			If value Is Nothing Then
				Return
			End If

			cbx_Items.SelectedValue = value
		End Set
	End Property

	Private Sub LoadItems(sender As Object, e As DoWorkEventArgs) Handles bw_LoadItems.DoWork
		RaiseEvent LoadBegin()
		'Dim items = db_Products.GetProducts()
		'_items.Clear()
		Try
			bsItems.Clear()
		Catch

		End Try

		For Each item In db_Products.GetProducts()
			'Dim var As DataTables.ItemsDataRow = CType(CType(bsItems.AddNew(), DataRowView).Row, DataTables.ItemsDataRow)
			'var.ItemArray = {item.Id, item.Name, item.Stock, item.Price, item.Available}
			'Console.WriteLine(var.ItemArray)
			bsItems.Add(item)
			'bsItems.Add(DataTables.ItemsDataRow())
			'_items.AddItemsRow(item.Id, item.Name, item.Stock, item.Price, item.Available)
		Next
	End Sub

	Private Sub ControlLoaded(sender As Object, e As EventArgs) Handles Me.Load
		'bsItems.DataSource = _items
	End Sub

	Private Sub ItemsLoaded(sender As Object, e As RunWorkerCompletedEventArgs) Handles bw_LoadItems.RunWorkerCompleted
		'cbx_Items.DataSource = bsItems
		'cbx_Items.DisplayMember = "ItemName"
		'cbx_Items.ValueMember = "ItemID"
		'cbx_Items.Refresh()
		bsItems.ResetBindings(False)
		RaiseEvent LoadEnd()
	End Sub

	Public Sub Reload()
		bw_LoadItems.RunWorkerAsync()
	End Sub

	Private Sub IndexChanged(sender As Object, e As EventArgs) Handles cbx_Items.SelectedIndexChanged

	End Sub
End Class

File: M3Tools/Controls/Products/ProductsComboBox.cs
using System;
using System.ComponentModel;
using Microsoft.VisualBasic.CompilerServices;

namespace SPPBC.M3Tools
{

	/// <summary>
	/// 
	/// </summary>
	public partial class ProductsComboBox
	{
		/// <summary>
		/// 
		/// </summary>
		public event LoadBeginEventHandler LoadBegin;

		/// <summary>
		/// 
		/// </summary>
		public delegate void LoadBeginEventHandler();
		/// <summary>
		/// 
		/// </summary>
		public event LoadEndEventHandler LoadEnd;

		/// <summary>
		/// 
		/// </summary>
		public delegate void LoadEndEventHandler();
		/// <summary>
		/// 
		/// </summary>
		public event SelectedItemChangedEventHandler SelectedItemChanged;

		/// <summary>
		/// 
		/// </summary>
		/// <param name="newValue"></param>
		public delegate void SelectedItemChangedEventHandler(int newValue);

		/// <summary>
		/// The currently selected inventory item
		/// </summary>
		public Types.Product SelectedItem
		{
			get
			{
				return cbx_Items.SelectedItem as Types.Product;
			}
		}

		/// <summary>
		/// The currently selected index
		/// </summary>
		public int SelectedIndex
		{
			get
			{
				return cbx_Items.SelectedIndex;
			}
			set
			{
				if (value < 0 || value > cbx_Items.Items.Count)
				{
					throw new ArgumentException();
				}

				cbx_Items.SelectedIndex = value;
			}
		}

		/// <summary>
		/// The currently selected value
		/// </summary>
		public object SelectedValue
		{
			get
			{
				return cbx_Items.SelectedValue;
			}
			set
			{
				if (value is null)
				{
					return;
				}

				cbx_Items.SelectedValue = value;
			}
		}

		/// <summary>
		/// 
		/// </summary>
		public ProductsComboBox()
		{
			InitializeComponent();
		}

		private void LoadItems(object sender, DoWorkEventArgs e)
		{
			LoadBegin?.Invoke();
			// Dim items = db_Products.GetProducts()
			// _items.Clear()
			try
			{
				bsItems.Clear();
			}
			catch
			{

			}

			foreach (var item in db_Products.GetProducts())
				// Dim var As DataTables.ItemsDataRow = CType(CType(bsItems.AddNew(), DataRowView).Row, DataTables.ItemsDataRow)
				// var.ItemArray = {item.Id, item.Name, item.Stock, item.Price, item.Available}
				// Console.WriteLine(var.ItemArray)
				// bsItems.Add(DataTables.ItemsDataRow())
				// _items.AddItemsRow(item.Id, item.Name, item.Stock, item.Price, item.Available)
				bsItems.Add(item);
		}

		private void ControlLoaded(object sender, EventArgs e)
		{
			// bsItems.DataSource = _items
		}

		private void ItemsLoaded(object sender, RunWorkerCompletedEventArgs e)
		{
			// cbx_Items.DataSource = bsItems
			// cbx_Items.DisplayMember = "ItemName"
			// cbx_Items.ValueMember = "ItemID"
			// cbx_Items.Refresh()
			bsItems.ResetBindings(false);
			LoadEnd?.Invoke();
		}

		/// <summary>
		/// Reloads the control
		/// </summary>
		public void Reload()
		{
			bw_LoadItems.RunWorkerAsync();
		}

		private void IndexChanged(object sender, EventArgs e)
		{
			SelectedItemChanged?.Invoke(Conversions.ToInteger(SelectedValue));
		}
	}
}
